zenoh/src/net/routing/interceptor/downsampling.rs

```diff
@@ -0,0 +1,160 @@
+//
+// Copyright (c) 2024 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+//! ⚠️ WARNING ⚠️
+//!
+//! This module is intended for Zenoh's internal use.
+//!
+//! [Click here for Zenoh's documentation](../zenoh/index.html)
+
+use crate::net::routing::interceptor::*;
+use std::sync::{Arc, Mutex};
+use zenoh_config::DownsamplingItemConf;
+use zenoh_protocol::core::key_expr::OwnedKeyExpr;
+
+const RATELIMIT_STRATEGY: &str = "ratelimit";
+
+// EgressMsgDownsamplerRatelimit implements default ratelimit strategy, it has following
+// configuration parameters:
+//   threshold  - mandatory, indicates the minimum interval between messages.
+//                All messages arriving with a smaller interval will be skipped
+//   keyexprs   - optional, list of key expressions, only messages matching this expression will be processed,
+//                the rest will be passed as is.
+//   interfaces - optional, list of interfaces messages will be processed on,
+//                the rest will be passed as is.
+//
+// TODO: we can also add decimation strategy, with mandatory "factor" parameter
+
+pub(crate) struct EgressMsgDownsamplerRatelimit {
+    keyexprs: Option<Vec<OwnedKeyExpr>>,
+    threshold: std::time::Duration,
+    latest_message_timestamp: Arc<Mutex<std::time::Instant>>,
```

Comment:
I wonder wether we could use [AtomicInstant](https://docs.rs/atomic-instant/latest/atomic_instant/struct.AtomicInstant.html) in this case.

Reply:
Or we could use [this one](https://docs.rs/async-liveliness-monitor/0.1.1/async_liveliness_monitor/support/struct.AtomicInstant.html), it has way better resolution (Me? Shimming my own impls? Whatever do you mean?)

Reply:
I understand that I am poisoned by C/C++, but adding a new dependency for the sake of such a small improvement.. How normal is this?
(It seems that none of these crates have been used in zenoh yet)

Reply:
Super! On some of my past projects, the introduction of a new dependency required approval from the CTO and the legal department 😆